src/pip/_internal/main.py

```diff
@@ -19,6 +19,24 @@
 logger = logging.getLogger(__name__)
 
 
+# Do not run this directly! Running pip in-process is unsupported and
+# unsafe.
+#
+# Also, the location of this function may change, so calling it directly
+# is not portable across different pip versions.  If you have to call
+# this function, and understand and accept the implications of doing so,
+# the best approach is to use runpy as follows:
```

Comment:
Suggested rephrasing (no need to add me as a author for this in commit history):

> Do not import and use main() directly! The name and/or location of this function is subject to change, so using it directly is not supported and not portable across different pip versions.
> 
> Running pip in-process is unsupported and unsafe. This is elaborated in https://pip.pypa.io/en/stable/user_guide/#using-pip-from-your-program, which also provides suggestions that should work for nearly all users that are considering importing and using main() directly.
> 
> However, we know that certain users will still want to invoke pip in-process. If you understand and accept the implications of using pip in an unsupported manner, the best approach is to use runpy to avoid depending on the exact location of this entry point.

Reply:
Does this create a new process? Or uses the current, hence why destroying it once done?

Reply:
`runpy` will invoke `pip.__main__` as `__name__ == "__main__"` like `python -m pip` would, which does exit the process when done.

Plus the fact that this cements the idea that don't do things after calling pip's main is good too IMO.

Reply:
@pradyunsg Yes, I like your words a lot better. I'll incorporate them, thanks.

@gaborbernat No, runpy doesn't create a new process. It exits once done, because that's how runpy works (`src/pip/__main__.py` exits the process on completion, and runpy just executes that script).

Reply:
This limits significantly the usefulness of this 🤔 The main reason one would want to call pip in-process is to save the process creation time (non-negligable on Windows, especially when you have many interactions to perform with pip). Can one catch the SystemExit to stop exiting after invocation?

Reply:
One can do whatever one likes 🙂 This whole thing is utterly unsupported, so feel free to try whatever you want. Even using runpy isn't supported, it's just a way of having your code break less often (when we moved the location of `main` recently, a lot of code broke - that code might have survived the move if it had been using runpy).

Personally, and totally off the record, I'm unclear whether I consider catching `SystemExit` here a cool hack or a horrifying abuse 😉

Reply:
> Can one catch the SystemExit to stop exiting after invocation?

Yes.

Reply:
> cool hack or a horrifying abuse

Both. 🙂